libs/s25main/gameTypes/CompressedData.cpp

```diff
@@ -33,61 +34,82 @@ bool CompressedData::DecompressToFile(const boost::filesystem::path& filePath, u
         return false;
     }
 
-    auto uncompressedData = std::unique_ptr<char[]>(new char[length]);
-
-    unsigned outLength = length;
-
-    int err = BZ2_bzBuffToBuffDecompress(uncompressedData.get(), &outLength, &data[0], data.size(), 0, 0);
-    if(err != BZ_OK)
+    try
     {
-        LOG.write("FATAL ERROR: BZ2_bzBuffToBuffDecompress failed with code %d\n") % err;
-        return false;
-    }
+        const auto uncompressedData = decompress(data, uncompressedLength);
 
-    if(outLength != length)
-    {
-        LOG.write("FATAL ERROR: Length mismatch after decompressing. Expected: %u, got %u\n") % length % outLength;
-        return false;
-    }
+        if(!file.write(uncompressedData.data(), uncompressedData.size()))
+        {
+            LOG.write("FATAL ERROR: Writing to %s failed\n") % filePath;
+            return false;
+        }
 
-    if(!file.write(uncompressedData.get(), length))
+        if(checksum)
+            *checksum = CalcChecksumOfBuffer(uncompressedData);
+    } catch(const std::runtime_error& err)
     {
-        LOG.write("FATAL ERROR: Writing to %s failed\n") % filePath;
+        LOG.write("FATAL ERROR: 1%\n") % err.what();
         return false;
     }
 
-    if(checksum)
-        *checksum = CalcChecksumOfBuffer(uncompressedData.get(), length);
-
     return true;
 }
 
 bool CompressedData::CompressFromFile(const boost::filesystem::path& filePath, unsigned* checksum /* = nullptr */)
 {
     boost::nowide::ifstream file(filePath, std::ios::binary | std::ios::ate);
-    length = static_cast<unsigned>(file.tellg());
-    data.resize(static_cast<int>(std::ceil(length * 1.1))
-                + 600); // Buffer should be at most 1% bigger + 600 Bytes according to docu
+    uncompressedLength = static_cast<unsigned>(file.tellg());
     file.seekg(0);
 
-    auto uncompressedData = std::unique_ptr<char[]>(new char[length]);
-
-    if(!file.read(uncompressedData.get(), length))
+    std::vector<char> uncompressedData(uncompressedLength);
+    if(!file.read(uncompressedData.data(), uncompressedLength))
     {
         LOG.write("Could not read from %s\n") % filePath;
         return false;
     }
 
-    unsigned compressedLen = data.size();
-    int err = BZ2_bzBuffToBuffCompress(&data[0], &compressedLen, uncompressedData.get(), length, 9, 0, 250);
-    if(err != BZ_OK)
+    try
+    {
+        data = compress(uncompressedData);
+    } catch(const std::runtime_error& err)
     {
-        LOG.write("FATAL ERROR: BZ2_bzBuffToBuffCompress failed with error: %d\n") % err;
+        LOG.write("FATAL ERROR: 1%\n") % err.what();
```

Comment:
isnt there a % missing before the 1?

Reply:
Yes, good catch!
It is practically impossible to trigger this codepath though, so can't be tested: http://linux.math.tifr.res.in/manuals/html/manual_3.html#SEC36